---
title: "Single-nuc QC & filtering workflow"
author: "Carla Cohen"
date: "`r Sys.Date()`"
output: html_document
---

# QC-Filter workflow

This workflow follows the scflow quantnuclei pipeline. The overall aim is to calculate QC metrics and perform basic filtering using MADs and manual thresholds. 
This specific script analyses the complete Achilles dataset from three sequencing batches:
2021.10.11
2022.08.08
2023.08.22

The working directory is  
/project/tendonhca/ccohen/chromium/analysis/20230118_achilles/  

The inputs of the workflow are:
* count matrices generated by scflow quantnuclei pipeline (kallisto outputs)
* qc_filter.yml to specify parameters
* metadata.txt file to provide metadata

The steps of the workflow are:  

1.  Set up & use raw sce object  
* load packages, read in yml and metadata files, create output directory  
* import counts matrices & create sce objects  
* save the raw sce  
* calculate and plot barcode ranks
* calculate QC metrics (nCounts/sum, nFeatures/detected)  

2.  Perform minimal filtering  
* create sce_filter_minimal with sum > 3 and detected > 20  
* recalculate the QC metrics (nCounts/sum, nFeatures/detected, % mito) on sce_filter_minimal  
* convert EnsemblIDs to gene names 
* add patient metadata to the object
* plot number of cells before & after minimal filtering  
* calculate the MADs (statistical filtering thresholds)
* save the unfiltered sce
* plot the QC metrics showing the MADs  

3. Perform MAD filtering  
* filter based on MADs
* plot number of cells after MAD filtering  
* plot the QC metrics after MAD filtering  

4. Perform manual filtering  
* filter across all samples with manual thresholds set in the yml file
* plot number of cells after manual filtering  
* plot the QC metrics after manual filtering  
* filter individual samples further as required and plot again
* save the filtered sce
* convert sce to Seurat object and save 

The outputs will be saved in a directory in the format "date_QC-Filter.dir"  
This will contain subdirectories for plots and RDS objects  
The filtering parameters are saved in the file  "manual_filters.txt"  
A summary of the run is included at the end of the Rmd document    


Plot dimensions - these may need to be updated according to the number of datasets analysed  
These dimensions work well for about 10 samples  
This could be parameterised  

Barcode ranks 10 x 20  
Cell number before & after 10 x 20 
Cell number 10 x 6  
QC metrics 12 x 14  
UMI vs genes 10 x 20  
Novelty 10 x 10  
nUmi Vln 10 x 12  
% mt Vln 10 x 12  



## 1. Set up and use raw sce object 



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```


#### Load packages, read in yml and metadata files, create output directory 

```{r, include=FALSE}

library(yaml)
library(tidyverse)
library(BUSpaRse)
library(biomaRt)
library(cowplot)
library(scater)
library(viridis)
library(Seurat)

# Read in the yml file
ini <- read_yaml("qc_filter.yml")

#Read in the metadata file
metadata <- read.table("Files.dir/metadata.txt", sep = "\t", header = TRUE)

# make a new output folder for each run, with the date & time in the directory name
date <- Sys.Date() %>% str_replace_all("-", "")
time <- format(Sys.time(), "%X") %>% 
str_replace_all(":", "-") %>%
    str_sub(1,5)
directory <- paste0(date,"_", time, "_QC-Filter.dir")
dir.create(directory, showWarnings = FALSE)

```

```{r}

# Print the parameters used
ini

```

####  Read in the data and create the Single Cell Experiment Object

Create the sce object with zero filtering and save it. 

```{r, echo=FALSE, warning=FALSE}
#make a list of sample files
file_list <- list.files("kallisto.dir/", pattern="genes.mtx", recursive = TRUE)
names(file_list) <- str_replace(file_list, "/bus/genecount/genes.mtx", "") %>% 
     str_remove("-[:alpha:]+-date[:digit:].+")

#Make a new list for the seurat objects
sce <- list()

dir.create(paste0(directory, "/RDS_objects.dir/"))
dir.create(paste0(directory, "/RDS_objects.dir/raw/"))

# generate a list of sce objects
for (i in 1:length(file_list)){
    
    name <- str_replace(file_list[i], "genes.mtx", "")
    
    res_mat <- read_count_output(dir = paste0("kallisto.dir/",name),
                                 name = "genes", 
                                 tcc = FALSE)
    
    project_title <- str_replace(file_list[i], "/bus/genecount/genes.mtx", "") %>% 
        str_remove("-[:alpha:]+-date[:digit:].+")
    
    sce[[i]] <- SingleCellExperiment(list (counts = res_mat), 
                             mainExpName = project_title, 
        )

    colData(sce[[i]])$sample <- as.character(project_title)
    
    saveRDS(sce[[i]], 
            paste0(directory, "/RDS_objects.dir/raw/",
                   mainExpName(sce[[i]]),"_raw_SingleCellExp.rds"))

}


```

Add names to the list
```{r}
names(sce) <- str_replace(file_list, "/bus/genecount/genes.mtx", "") %>% 
        str_remove("-[:alpha:]+-date[:digit:].+")
sce
```


#### Calculate and plot barcode ranks  


```{r}
#apply barcodeRanks to all sce
br.out <- lapply(sce, DropletUtils::barcodeRanks)
```


```{r, fig.width=14, fig.height=24, warning = FALSE}
plot_list <- list()

for (i in 1:length(sce)){
    barcode_data = as.data.frame(br.out[[i]])

    barcode_points = data.frame(
        type = c("inflection", "knee"),
        value = c(br.out[[i]]@metadata$inflection, br.out[[i]]@metadata$knee))


    plot_list[[i]] <- ggplot(data = barcode_data, aes(x = rank, y = total)) +
          geom_point() +
          geom_hline(data = barcode_points,
                     aes(yintercept = value,
                         colour = type), linetype = 2) +
          scale_x_log10() +
          scale_y_log10() +
          labs(title = mainExpName(sce[[i]]),
               x = "Log Rank",
               y = "Log counts")
}


#add an overall title and arrange the graphs on one page
title <- ggdraw() + draw_label("Barcode rank plots", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 24, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))

p

```



```{r}
dir.create(paste0(directory, "/QC_figures"), showWarnings = FALSE)
ggsave(paste0(directory, "/QC_figures/Barcode_rank.", ini$file_type), 
device = ini$file_type, width = 14, height = 24, bg = "white")
```


#### Calculate QC metrics 

* number of counts per droplet (sum/nCounts)  
* number of genes per droplet (detected/nFeatures)  
* % mitochondrial reads per droplet  
* Number of UMIs per gene per droplet  


Identify EnsemblIDs for mitochondrial genes

```{r, echo=FALSE}

# Access the useMart database
# useMart keeps not working every so often (issue https://github.com/grimbough/biomaRt/issues/31)
ensembl <- try(useMart("ensembl", dataset = "hsapiens_gene_ensembl"))
if(class(ensembl) == "try-error"){
  httr::set_config(httr::config(ssl_verifypeer = FALSE))
  ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
}

# get a df of ensemble_id vs gene symbol
mapping <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"), mart = ensembl)

#identify mitochondrial genes starting MT-
mito <- mapping %>% filter(grepl("^MT-", hgnc_symbol))

```

Save the mapping object for use in future pipelines

```{r, include = FALSE}
dir.create("Files.dir", showWarnings = FALSE)
write.table(mapping, "Files.dir/mapping.txt", quote = FALSE, sep = "\t", row.names = FALSE)
```


## 2. Perform minimal filtering  

#### Create sce_filter_minimal with sum > 3 and detected > 20  
#### Recalculate the QC metrics (nCounts/sum, nFeatures/detected, % mito) on sce_filter_minimal
  

```{r}

sce_filter_minimal <- list()

for (i in 1:length(sce)) {
    
    # add basic QC metrics to the raw sce
    sce[[i]] <- scuttle::addPerCellQCMetrics(sce[[i]])
    
    # create a new object with minimal filtering and remove previous QC metrics
    sce_filter_minimal[[i]] <- sce[[i]][, sce[[i]]$sum > 3 & 
                                               sce[[i]]$detected > 20]
    sce_filter_minimal[[i]]$sum <- NULL
    sce_filter_minimal[[i]]$detected <- NULL
    sce_filter_minimal[[i]]$total <- NULL
    
    
    # recalculate the QC metrics and add to colData
    sce_filter_minimal[[i]] <- scuttle::addPerCellQCMetrics(sce_filter_minimal[[i]],
                             subsets = list(mito = mito$ensembl_gene_id))
    
    # add log10GenesPerUMI to colData
    colData(sce_filter_minimal[[i]])$log10GenesPerUMI <- log10(colData(sce_filter_minimal[[i]])$detected) / log10(colData(sce_filter_minimal[[i]])$sum)

}

# add names to the list
names(sce_filter_minimal) <- names(sce)

sce_filter_minimal

```


#### Add gene symbols to the rowData/rownames 

Where no gene symbol is available, leave in the EnsemblID   
For Bioconductor SingleCellExperiment this can be added as rowData, and the rownames of the count matrix.  
NB For a Seurat object this has to be added for each assay.  


```{r, echo=FALSE, message = FALSE}

df <- list()
for(i in 1:length(sce_filter_minimal)){
    
    #make a df with ensemble_id and hgnc_symbol
    df[[i]] <- data.frame("ensembl_gene_id" = rownames(sce_filter_minimal[[i]])) %>% 
        left_join(mapping)
    
    #add a column of unique names, remove the "_" from ensembl_ids
    df[[i]]$unique_name <- uniquifyFeatureNames(df[[i]]$ensembl_gene_id, df[[i]]$hgnc_symbol)
    df[[i]]$unique_name <- gsub("^_", "", df[[i]]$unique_name)
    df[[i]] <- df[[i]] %>% distinct(ensembl_gene_id, .keep_all = TRUE) #remove duplicated rows
    
    #add ensembl_IDs to row names
    rownames(df[[i]]) <- rownames(sce_filter_minimal[[i]])
    
    #add to the meta.features in the RNA assay
    rowData(sce_filter_minimal[[i]]) <- df[[i]]
    
    # Update the assay rownames
    rownames(sce_filter_minimal[[i]]) <- rowData(sce_filter_minimal[[i]])$unique_name
    
}

```

#### Add patient metadata to sce  

Read in patient metadata for each sample including the following fields 
NB this will change according to the dataset being analysed  

Patient metadata  
- age  
- sex  
- ethnicity  
- surgical procedure  
- disease status  


Tissue & processing metadata  
- anatomical site
- affected side  
- microanatomical site  
- time to freezing  
- sequencing date



```{r, echo=FALSE}
# TO DO add PCR cycles to metadata


for (i in 1:length(sce_filter_minimal)){
    
    #cat ("analysing sample ", i)
    #cat("\n")
    sample_name <- str_sub(file_list[[i]], 1, 7)
    info <- metadata %>% filter(Patient == sample_name)

    sce_filter_minimal[[i]]$patient <- info$Patient
    sce_filter_minimal[[i]]$age <- info$Age
    sce_filter_minimal[[i]]$sex <- info$Sex
    sce_filter_minimal[[i]]$ethnicity <- info$Ethnicity
    sce_filter_minimal[[i]]$surgical_procedure <- info$Surgical.procedure
    sce_filter_minimal[[i]]$disease_status <- info$Disease.status
    sce_filter_minimal[[i]]$anatomical_site <- info$Anatomical.site
    sce_filter_minimal[[i]]$affected_side <- info$Affected.side
    sce_filter_minimal[[i]]$time_to_freezing <- info$Time.to.freezing
    sce_filter_minimal[[i]]$sequencing_date <- str_extract(file_list[[i]], "[:digit:]{8}")
    sce_filter_minimal[[i]]$microanatomical_site <- str_replace(names(file_list)[[i]], "MSK[:digit:]+-[:alpha:]+-", "") %>% 
        str_replace("[:digit:]+", "")
    
}

```



#### Plot number of cells before & after minimal filtering  


```{r, fig.width=10, fig.height=20}
#create a df of number of cells per sample
df <- data.frame()

for (i in 1:length(sce)) {
    sample <- mainExpName(sce[[i]])
    cells <- ncol(sce[[i]])
    cells_minimal <- ncol(sce_filter_minimal[[i]])
    df[i, 1] <- sample
    df[i, 2] <- cells
    df[i, 3] <- cells_minimal

}

colnames(df) <- c("sample", "cells", "cells_minimal")

df <- df %>% pivot_longer(2:3, names_to = "filter", values_to = "number_of_cells")

df$filter <- factor(df$filter, levels = c("cells", "cells_minimal"))

#plot the number of cells per sample as a bar chart
p <- ggplot(df, aes(x = filter, y = number_of_cells, fill = filter))+
    geom_col()+
    ggtitle("Number of cells before and after minimal filtering")+
    theme(plot.title = element_text(size = 24))+
    theme(axis.text.x = element_blank())+
    geom_text(aes(label=number_of_cells), position=position_dodge(width=0.9), vjust=-0.25)+
    facet_wrap("sample", ncol = 3)+
    scale_fill_viridis_d(begin = 0, end = 0.75)+
    theme_cowplot()+
    theme(legend.position = "none")+
    theme(axis.text.x = element_text(angle = 45, hjust=1))
p   

```

```{r}
ggsave(paste0(directory, "/QC_figures/Minimal_filter_cellnumber.", ini$file_type), 
    device = ini$file_type, width = 10, height = 20, bg = "white")
```


### Plot the QC metrics after minimal filtering 

Plot the number of counts per cell (nCount_RNA/sum) with low library size cells.  
Plot the number of features per cell (nFeature_RNA/detected) with low number of features.  
Plot the mitochondrial content (percent_mt/subsets_mito_percent) with high percent mt.  


```{r, fig.width=12, fig.height=14}
plot_filtered_qc <- function(sce_list, filter_method){
    df <- list()
    
    for (i in 1:length(sce_list)){
            
        df[[i]] <- data.frame(
            sce_list[[i]]$sample,
            colnames(sce_list[[i]]),
            sce_list[[i]]$sum,
            sce_list[[i]]$detected,
            sce_list[[i]]$subsets_mito_percent
        )
            
            colnames(df[[i]]) <- c("sample", "barcode", "nCount_RNA", "nFeature_RNA", "percent_mt")
    }
        
        # make concatenated df
        
        df_all <- plyr::ldply(df)
        
        #plot low library size
        p1 <- ggplot(df_all, aes(x = sample, y = nCount_RNA))+
            geom_violin(trim = FALSE, fill = "#440154", width = 1, position = position_dodge(0))+
            scale_y_log10() + 
            theme_cowplot()+
            theme(axis.text.x = element_text(angle = 45, hjust=1))+
            ggtitle("Library size")
            
        # plot low number of features
        p2 <- ggplot(df_all, aes(x = sample, y = nFeature_RNA))+
            geom_violin(trim = FALSE, fill = "#440154", width = 1, position = position_dodge(0))+
            scale_y_log10() + 
            theme_cowplot()+
            theme(axis.text.x = element_text(angle = 45, hjust=1))+
            ggtitle("Number of features")
        
        # plot mitochondrial content
        
        p3 <- ggplot(df_all, aes(x = sample, y = percent_mt))+
            geom_violin(trim = FALSE, fill = "#440154", width = 1, position = position_dodge(0))+
            theme_cowplot()+
            theme(axis.text.x = element_text(angle = 45, hjust=1))+
            ggtitle("Mitochondrial reads")
        
        
        #add an overall title and arrange the graphs on one page
        title <- ggdraw() + draw_label(paste0("QC after ", filter_method, " filtering"), fontface='bold', size = 16)
        p <- plot_grid(p1, p2, p3, ncol = 1)
        p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))
        
        p
        
}

plot_filtered_qc(sce_filter_minimal, "minimal")

```

```{r}
ggsave(paste0(directory, "/QC_figures/Minimal_Vlnplots.", ini$file_type), 
    device = ini$file_type, width = 12, height = 14, bg = "white")
```


#### Plot the relationship between number of mitochondrial reads and total counts

```{r, fig.width=16, fig.height=24}
plot_umi_vs_mito <- function (sce_list, filter_method){
    plot_list <- list()
    for (i in 1:length(sce_list)) {
      
        plot_list[[i]] <- colData(sce_list[[i]]) %>% 
            as_tibble() %>%
            ggplot(aes(x=sum, y=subsets_mito_percent )) + 
            geom_point(aes(colour = detected )) + 
            scale_colour_viridis()+
            geom_vline(xintercept = ini$sum_filter) +
            geom_hline(yintercept = ini$mito_filter)+
            scale_x_log10()+
            labs(x = "log(nUMIs)")+
            ggtitle(mainExpName(sce_list[[i]])) +
            theme(plot.title = element_text(size = 12))
    }
    
    
    title <- ggdraw() + draw_label(paste0("UMIs vs % mito after ", filter_method, " filtering"), fontface='bold', size = 16)
    p <- plot_grid(plotlist = plot_list, label_size = 12, ncol = 3)
    p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 2))
    
    p
}

p <- plot_umi_vs_mito(sce_filter_minimal, "minimal")
p

```


```{r}
ggsave(paste0(directory, "/QC_figures/Minimal_nUMI_vs_mito.", ini$file_type), 
    device = ini$file_type, width = 16, height = 24, bg = "white")
```


#### Calculation of MADs

We will use statistical filtering to remove cells with low numbers of counts and features. 

Calculate QC metrics using perCellQCFilters which calculates outliers > 3 MADs from median of the data  
- log number of counts  
- log number of features  
- number of mitochondrial reads  

Adds these metrics to the colData.  
The discard column contains the final call for whether a particular cell should be discarded.

Save the raw single cell object that now includes metadata and calculations of QC metrics. 


```{r, warning=FALSE, echo=FALSE}
dir.create(paste0(directory, "/RDS_objects.dir/"))
dir.create(paste0(directory, "/RDS_objects.dir/unfiltered/"))
for (i in 1:length(sce)){
    QC_filters <- perCellQCFilters(sce_filter_minimal[[i]], 
                      sum.field = "sum", 
                      detected.field = "detected",
                      sub.fields = "subsets_mito_percent", 
                      nmads = 3, 
                      )
    colData(sce_filter_minimal[[i]]) <- cbind(colData(sce_filter_minimal[[i]]), QC_filters)
    
    saveRDS(sce[[i]], 
            paste0(directory, "/RDS_objects.dir/unfiltered/", mainExpName(sce[[i]]), "_unfiltered_SingleCellExp.rds"))
   
}

# print one example of the colData
# head(colData(sce_filter_minimal[[1]]))
```

#### Plot the QC metrics after minimal filtering showing the MADs


```{r, fig.width=12, fig.height=14, warning=FALSE}

plot_all_qc <- function (sce_list){
    df <- list()

    for (i in 1:length(sce_list)){
        df[[i]] <- data.frame(
            sce_list[[i]]$sample,
            colnames(sce_list[[i]]),
            sce_list[[i]]$sum,
            sce_list[[i]]$detected,
            sce_list[[i]]$low_lib_size,
            sce_list[[i]]$low_n_features, 
            sce_list[[i]]$subsets_mito_percent,
            sce_list[[i]]$high_subsets_mito_percent
        )
        
        colnames(df[[i]]) <- c("sample", "barcode", "nCount_RNA", "nFeature_RNA", "low_lib_size", "low_n_features", "percent_mt", "high_percent_mt")
    }
    
    # make concatenated df
    
    df_all <- plyr::ldply(df)
    
    #plot low library size
    p1 <- ggplot(df_all, aes(x = sample, y = nCount_RNA, fill = low_lib_size))+
        geom_violin(trim = FALSE, width = 3, position = position_dodge(0))+
        scale_y_log10() + 
        scale_fill_viridis_d(begin = 0, end = 0.75)+
        theme_cowplot()+
        theme(axis.text.x = element_text(angle = 45, hjust=1))+
        ggtitle("Statistical filtering: Low library size")
        
    # plot low number of features
    p2 <- ggplot(df_all, aes(x = sample, y = nFeature_RNA, fill = low_n_features))+
        geom_violin(trim = FALSE, width = 2, position = position_dodge(0))+
        scale_y_log10() + 
        scale_fill_viridis_d(begin = 0, end = 0.75)+
        theme_cowplot()+
        theme(axis.text.x = element_text(angle = 45, hjust=1))+
        ggtitle("Statistical filtering: Low number of features")
    
    # plot mitochondrial content
    
    p3 <- ggplot(df_all, aes(x = sample, y = percent_mt, fill = high_percent_mt))+
        geom_violin(trim = FALSE, width = 2, position = position_dodge(0))+
        scale_fill_viridis_d(begin = 0, end = 0.75)+
        theme_cowplot()+
        theme(axis.text.x = element_text(angle = 45, hjust=1))+
        ggtitle("Statistical filtering: Mitochondrial reads")
    
    
    #add an overall title and arrange the graphs on one page
    title <- ggdraw() + draw_label("Calculated statistical metrics", fontface='bold', size = 16)
    p <- plot_grid(p1, p2, p3, ncol = 1)
    p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))
    
    p
}

plot_all_qc(sce_filter_minimal)

```

```{r}
ggsave(paste0(directory, "/QC_figures/Minimal_Vlnplots_MADs.", ini$file_type), 
    device = ini$file_type, width = 12, height = 14, bg = "white")
```


## 3. Perform MAD filtering 

We will now filter the sce using the statistical MAD filtering method.  
Create an object removing the "discard" droplets (sce_filter_MAD).  
Note: An important thing to note is that, now that we have filtered this object, some of the QC metrics that were calculated across all genes (for colData) and across all cells (for rowData) are no longer correct for the filtered data set. We need to remove them, and if necessary recalculate them.


```{r}

sce_filter_MAD <- list()

for (i in 1:length(sce_filter_minimal)){
    
    # create the sce with discarded cells filtered out
    sce_filter_MAD[[i]] <- sce_filter_minimal[[i]][, which(sce_filter_minimal[[i]]$discard == FALSE)]
    
    # remove columns from metadata
    sce_filter_MAD[[i]]$low_lib_size <- NULL
    sce_filter_MAD[[i]]$low_n_features <- NULL
    sce_filter_MAD[[i]]$high_subsets_mito_percent <- NULL
    sce_filter_MAD[[i]]$discard <- NULL
    
}

# add names to the list
names(sce_filter_MAD) <- names(sce)

```

#### Plot number of cells after MAD filtering 

```{r, fig.width=10, fig.height=6}
plot_cell_number <- function(sce_list, cell_drop, filter_method){
    #create a df of number of cells or droplets per sample
df <- data.frame()

    for (i in 1:length(sce_list)) {
        sample <- mainExpName(sce_list[[i]])
        droplets <- ncol(sce_list[[i]])
        df[i,1] <- sample
        df[i, 2] <- droplets
}

    colnames(df) <- c("sample", "droplets")

#plot the number of cells per sample as a bar chart
    p <- ggplot(df, aes(x = sample, y = droplets, fill = sample))+
    geom_col()+
        ggtitle(paste0("Number of ", cell_drop, " per sample after ", filter_method, " filtering"))+
        geom_text(aes(label=droplets), position=position_dodge(width=0.9), vjust=-0.25)+
    theme_cowplot()+
        theme(plot.title = element_text(size = 20))+
        theme(axis.text.x = element_text(size = 12, angle = 45, hjust=1))+
        theme(legend.position="none")+
        scale_fill_viridis_d()
    
p   
}
    
plot_cell_number(sce_filter_MAD, "cells", "MAD")
    
```

```{r}
ggsave(paste0(directory, "/QC_figures/MAD_filter_cellnumber.", ini$file_type), 
    device = ini$file_type, width = 10, height = 6, bg = "white")
```


#### Plot the QC metrics after MAD filtering  
```{r, fig.width=12, fig.height=14, warning=FALSE}

plot_filtered_qc(sce_filter_MAD, "statistical")
```

```{r}
ggsave(paste0(directory, "/QC_figures/MAD_filter_Vlnplots.", ini$file_type), 
    device = ini$file_type, width = 12, height = 14, bg = "white")
```

#### Plot the relationship between nUMI and % MT after MAD filtering 

```{r, fig.width=10, fig.height=20}
p <- plot_umi_vs_mito(sce_filter_MAD, "MAD")
p
```

```{r}
ggsave(paste0(directory, "/QC_figures/MAD_filter_nUMI_vs_mito.", ini$file_type), 
    device = ini$file_type, width = 10, height = 20, bg = "white")

```


#### Plot the UMI vs features coloured by % MT after MAD filtering 

```{r, fig.width=14, fig.height=24, message = FALSE}

plot_qc <- function (sce_list, filter_method){
    plot_list <- list()
    for (i in 1:length(sce_list)) {
      
        plot_list[[i]] <- colData(sce_list[[i]]) %>% 
            as_tibble() %>%
            ggplot(aes(x=sum, y=detected)) + 
            geom_point(aes(colour = subsets_mito_percent)) + 
            scale_colour_viridis()+
            stat_smooth(method=lm) +
            scale_x_log10() + 
            scale_y_log10() + 
            geom_vline(xintercept = ini$sum_filter) +
            geom_hline(yintercept = ini$detected_filter)+
            ggtitle(mainExpName(sce_list[[i]])) +
            theme(plot.title = element_text(size = 12))
    }
    
    
    title <- ggdraw() + draw_label(paste0("UMIs vs genes after ", filter_method, " filtering"), fontface='bold', size = 24)
    p <- plot_grid(plotlist = plot_list, label_size = 12, ncol = 3)
    p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 2))
    
    p
}



plot_qc(sce_filter_MAD, "statistical")


```

```{r}
ggsave(paste0(directory, "/QC_figures/MAD_filter_UMIvsgenes.", ini$file_type), 
    device = ini$file_type, width = 14, height = 24, bg = "white")
```



#### Plot the Novelty after MAD filtering   

Visualize the overall novelty of the gene expression by visualizing the genes detected per UMI

We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturating the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric. Generally, we expect the novelty score to be above 0.80.

```{r, echo=FALSE, fig.width = 10, fig.height=16, warning=FALSE}

plot_novelty <- function (sce_list, filter_method){
    plot_list <- list()
    for (i in 1:length(sce_list)) {
      
        plot_list[[i]] <- colData(sce_list[[i]]) %>%
            as_tibble()%>%
            ggplot(aes(x=log10GenesPerUMI, fill = sample)) +
            geom_density() +
            scale_fill_viridis_d()+
            ggtitle(mainExpName(sce_list[[i]]))+
            theme(plot.title = element_text(size = 12))+
            xlim(0.5, 1)+
            theme_cowplot()+
            theme(legend.position="none")
            
    }
    
    title <- ggdraw() + draw_label(paste0("Novelty after ", filter_method, " filtering"), fontface='bold', size = 24)
    p <- plot_grid(plotlist = plot_list, label_size = 12, ncol = 3)
    p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))
    
    p
}

plot_novelty(sce_filter_MAD, "statistical")

```

```{r}
ggsave(paste0(directory, "/QC_figures/MAD_filter_Novelty.", ini$file_type), 
    device = ini$file_type, width = 10, height = 14, bg = "white")
```

### 4. Perform filtering with manual thresholds  


#### Manual threshold filtering  

- Filter cells, keeping those with more than 300 UMI, less than 5% mitochondrial UMI, and more than 300 genes detected. 
- These thresholds can be updated in the yml file. 

First have another look at the % mt plot

```{r, fig.width=10, fig.height=16, warning=FALSE, message = FALSE}
plot_mito <- function (sce_list){
    plot_list <- list()

for (i in 1:length(sce_list)) {

    # do a violin plot
    plot_list[[i]] <- plotColData(sce_list[[i]], x = "sample", y = "subsets_mito_percent", colour = "sample")+
        #add a title
        ggtitle(mainExpName(sce_list[[i]])) +
        theme(plot.title = element_text(size = 12),
              axis.title.x = element_blank(),
              legend.position = "none")+
        scale_colour_viridis_d()
     }
    
#add an overall title and arrange the graphs on one page
title <- ggdraw() + draw_label("% mitochondrial reads per droplet", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 24, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))
    
    p
}

plot_mito(sce_filter_MAD)

```

```{r}
ggsave(paste0(directory, "/QC_figures/MAD_filter_mito_Vln.", ini$file_type), 
    device = ini$file_type, width = 10, height = 16, bg = "white")
```

#### Perform the filtering  

```{r}
sce_filter_threshold <- list()

for (i in 1:length(sce_filter_MAD)) {
    sce_filter_threshold[[i]] <- sce_filter_MAD[[i]][, sce_filter_MAD[[i]]$sum > ini$sum_filter & 
                                               sce_filter_MAD[[i]]$subsets_mito_percent < ini$mito_filter & 
                                               sce_filter_MAD[[i]]$detected > ini$detected_filter]
}

names(sce_filter_threshold) <- names(sce_filter_MAD)
```

#### Plot number of cells after threshold filtering  


```{r, fig.width=10, fig.height=6}


plot_cell_number(sce_filter_threshold, "cells", "threshold")

```

```{r}
ggsave(paste0(directory, "/QC_figures/Threshold_filter_cellnumber.", ini$file_type), 
    device = ini$file_type, width = 10, height = 6, bg = "white")
```

#### Plot QC metrics after manual threshold filtering

```{r, fig.width=12, fig.height=14}
plot_filtered_qc(sce_filter_threshold, "threshold")
```

```{r}
ggsave(paste0(directory, "/QC_figures/Threshold_filter_Vlnplots.", ini$file_type), 
    device = ini$file_type, width = 10, height = 6, bg = "white")
```

#### Plot the relationship between nUMI and % MT after manual threshold filtering 

```{r, fig.width=10, fig.height=24}
p <- plot_umi_vs_mito(sce_filter_threshold, "threshold")
p
```

```{r}
ggsave(paste0(directory, "/QC_figures/Threshold_filter_nUMI_vs_mito.", ini$file_type), 
    device = ini$file_type, width = 10, height = 24, bg = "white")

```

#### #### Plot the UMI vs features coloured by % MT after threshold filtering

```{r, fig.height=24, fig.width=14, message = FALSE}
plot_qc(sce_filter_threshold, "threshold")

```
```{r}
ggsave(paste0(directory, "/QC_figures/Threshold_filter_UMIvsgenes.", ini$file_type), 
    device = ini$file_type, width = 14, height = 24, bg = "white")
```

#### Additional manual filtering on per-sample basis

From viewing the above plots, decide if any additional filtering is needed. 
This will vary according to the data set being analysed.  
Starting with sum/nCounts per droplet. 

```{r, fig.width=10, fig.height=16, message=FALSE}
plot_nUMI <- function (sce_list){
    plot_list <- list()

for (i in 1:length(sce_list)) {
    
    # do a violin plot
    plot_list[[i]] <- plotColData(sce_list[[i]], x = "sample", y = "sum", colour = "sample")+
        #add a title
        ggtitle(mainExpName(sce_list[[i]])) +
        theme(plot.title = element_text(size = 12),
              axis.title.x = element_blank(),
              legend.position = "none")+
        scale_y_log10()+
        scale_colour_viridis_d()
}

#add an overall title and arrange the graphs on one page
title <- ggdraw() + draw_label("UMI counts per droplet", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 24, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))

p
}

plot_nUMI(sce_filter_threshold)

```

```{r}
ggsave(paste0(directory, "/QC_figures/Threshold_filter_nCounts_Vln.", ini$file_type), 
    device = ini$file_type, width = 10, height = 16, bg = "white")
```

Perform additional manual filtering of UMIs & create a df with the filtering thresholds.  

Increase filtering of the following samples
MSK0785-Ach-MTJ 	nUMI 1000
MSK1250-Ach-Enth	nUMI 500
MSK1250-Ach-MB2	nUMI 1000
MSK1250-Ach-MTJ	nUMI 500
MSK1556-Ach-Enth	nUMI 800
MSK1156-Ach-MB	nUMI 800
MSK1556-Ach-MTJ	nUMI 1000
MSK1556-Ach-muscle	nUMI1000
MSK1687-Ach-Enth	nUMI 500
MSK1687-Ach-MB	nUMI 800
MSK1687-Ach-MTJ	nUMI 800
MSK1681-Ach_Enth	nUMI 500
MSK1691-Ach-MB	nUMI 800
MSK1691-Ach-MTJ	nUMI 800


```{r}

sce_filter_manual <- sce_filter_threshold
sce_filter_manual[["MSK0785-Ach-Enth"]] <- 
    sce_filter_threshold[["MSK0785-Ach-Enth"]][, sce_filter_threshold[["MSK0785-Ach-Enth"]]$sum > 1000 ]
sce_filter_manual[["MSK0785-Ach-MB"]] <- 
    sce_filter_threshold[["MSK0785-Ach-MB"]][, sce_filter_threshold[["MSK0785-Ach-MB"]]$sum > 500 ]
sce_filter_manual[["MSK0785-Ach-MTJ"]] <- 
    sce_filter_threshold[["MSK0785-Ach-MTJ"]][, sce_filter_threshold[["MSK0785-Ach-MTJ"]]$sum > 1000 ]
sce_filter_manual[["MSK0785-Ach-muscle"]] <- 
    sce_filter_threshold[["MSK0785-Ach-muscle"]][, sce_filter_threshold[["MSK0785-Ach-muscle"]]$sum > 1000 ]

sce_filter_manual[["MSK1250-Ach-Enth"]] <- 
    sce_filter_threshold[["MSK1250-Ach-Enth"]][, sce_filter_threshold[["MSK1250-Ach-Enth"]]$sum > 500 ]
sce_filter_manual[["MSK1250-Ach-MB2"]] <- 
    sce_filter_threshold[["MSK1250-Ach-MB2"]][, sce_filter_threshold[["MSK1250-Ach-MB2"]]$sum > 1000 ]
sce_filter_manual[["MSK1250-Ach-MTJ"]] <- 
    sce_filter_threshold[["MSK1250-Ach-MTJ"]][, sce_filter_threshold[["MSK1250-Ach-MTJ"]]$sum > 1000 ]
sce_filter_manual[["MSK1250-ACH-muscle"]] <- 
    sce_filter_threshold[["MSK1250-ACH-muscle"]][, sce_filter_threshold[["MSK1250-ACH-muscle"]]$sum > 1000 ]

sce_filter_manual[["MSK1284-Ach-Enth"]] <- 
    sce_filter_threshold[["MSK1284-Ach-Enth"]][, sce_filter_threshold[["MSK1284-Ach-Enth"]]$sum > 500 ]

sce_filter_manual[["MSK1556-Ach-Enth"]] <- 
    sce_filter_threshold[["MSK1556-Ach-Enth"]][, sce_filter_threshold[["MSK1556-Ach-Enth"]]$sum > 800 ]
sce_filter_manual[["MSK1556-Ach-MB"]] <- 
    sce_filter_threshold[["MSK1556-Ach-MB"]][, sce_filter_threshold[["MSK1556-Ach-MB"]]$sum > 800 ]
sce_filter_manual[["MSK1556-Ach-MTJ"]] <- 
    sce_filter_threshold[["MSK1556-Ach-MTJ"]][, sce_filter_threshold[["MSK1556-Ach-MTJ"]]$sum > 1000 ]
sce_filter_manual[["MSK1556-ACH-muscle"]] <- 
    sce_filter_threshold[["MSK1556-ACH-muscle"]][, sce_filter_threshold[["MSK1556-ACH-muscle"]]$sum > 1000 ]

sce_filter_manual[["MSK1687-ACH-Enth"]] <- 
    sce_filter_threshold[["MSK1687-ACH-Enth"]][, sce_filter_threshold[["MSK1687-ACH-Enth"]]$sum > 500 ]
sce_filter_manual[["MSK1687-ACH-MB"]] <- 
    sce_filter_threshold[["MSK1687-ACH-MB"]][, sce_filter_threshold[["MSK1687-ACH-MB"]]$sum > 800 ]
sce_filter_manual[["MSK1687-ACH-MTJ"]] <- 
    sce_filter_threshold[["MSK1687-ACH-MTJ"]][, sce_filter_threshold[["MSK1687-ACH-MTJ"]]$sum > 800 ]

sce_filter_manual[["MSK1691-ACH-Enth"]] <- 
    sce_filter_threshold[["MSK1691-ACH-Enth"]][, sce_filter_threshold[["MSK1691-ACH-Enth"]]$sum > 500 ]
sce_filter_manual[["MSK1691-ACH-MB"]] <- 
    sce_filter_threshold[["MSK1691-ACH-MB"]][, sce_filter_threshold[["MSK1691-ACH-MB"]]$sum > 800 ]
sce_filter_manual[["MSK1691-ACH-MTJ"]] <- 
    sce_filter_threshold[["MSK1691-ACH-MTJ"]][, sce_filter_threshold[["MSK1691-ACH-MTJ"]]$sum > 800 ]

# create a df of the filtering parameters used
manual_filters <- data.frame(
    "sample" = names(sce_filter_manual), 
    "counts_filter" = rep(ini$detected_filter, length(sce_filter_manual)),
    "features_filter" = rep(ini$detected_filter, length(sce_filter_manual)),
    "mito_filter" = rep(ini$mito_filter, length(sce_filter_manual)))

# update the manual_filters df  

manual_filters[manual_filters$sample =="MSK0785-Ach-Enth", "counts_filter"] <- 1000
manual_filters[manual_filters$sample =="MSK0785-Ach-MB", "counts_filter"] <- 500
manual_filters[manual_filters$sample =="MSK0785-Ach-MTJ", "counts_filter"] <- 1000
manual_filters[manual_filters$sample =="MSK0785-Ach-muscle", "counts_filter"] <- 1000

manual_filters[manual_filters$sample =="MSK1250-Ach-Enth", "counts_filter"] <- 500
manual_filters[manual_filters$sample =="MSK1250-Ach-MB2", "counts_filter"] <- 1000
manual_filters[manual_filters$sample =="MSK1250-Ach-MTJ", "counts_filter"] <- 1000
manual_filters[manual_filters$sample =="MSK1250-Ach-muscle", "counts_filter"] <- 1000

manual_filters[manual_filters$sample =="MSK1284-Ach-Enth", "counts_filter"] <- 500

manual_filters[manual_filters$sample =="MSK1556-Ach-Enth", "counts_filter"] <- 800
manual_filters[manual_filters$sample =="MSK1556-Ach-MB", "counts_filter"] <- 800
manual_filters[manual_filters$sample =="MSK1556-Ach-MTJ", "counts_filter"] <- 1000
manual_filters[manual_filters$sample =="MSK1556-Ach-muscle", "counts_filter"] <- 1000

manual_filters[manual_filters$sample =="MSK1687-ACH-Enth", "counts_filter"] <- 500
manual_filters[manual_filters$sample =="MSK1687-ACH-MB", "counts_filter"] <- 800
manual_filters[manual_filters$sample =="MSK1687-ACH-MTJ", "counts_filter"] <- 800

manual_filters[manual_filters$sample =="MSK1691-ACH-Enth", "counts_filter"] <- 500
manual_filters[manual_filters$sample =="MSK1691-ACH-MB", "counts_filter"] <- 800
manual_filters[manual_filters$sample =="MSK1691-ACH-MTJ", "counts_filter"] <- 800


```

#### Plot the number of cells after manual filtering

```{r, fig.width=10, fig.height = 6}
plot_cell_number(sce_filter_manual, "cell", "additional manual")
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter_cellnumber.", ini$file_type), 
       device = ini$file_type, width = 10, height = 6, bg = "white")
```


#### Plot the QC metrics after manual filtering  
```{r, fig.width=12, fig.height=14}
plot_filtered_qc(sce_filter_manual, "additional UMI")
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter_Vlnplots.", ini$file_type), 
device = ini$file_type, width = 12, height = 14, bg = "white")
```


#### Plot the relationship between nUMI and % MT after manual filtering 

```{r, fig.width=14, fig.height=24}
p <- plot_umi_vs_mito(sce_filter_manual, "manual")
p
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter_nUMI_vs_mito.", ini$file_type), 
    device = ini$file_type, width = 14, height = 24, bg = "white")

```



#### Plot the UMI vs genes detected after manual filtering

```{r, fig.width=14, fig.height=24, message = FALSE}
plot_qc(sce_filter_manual, "additional UMI")
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter_UMIvsgenes.", ini$file_type), 
device = ini$file_type, width = 10, height = 20, bg = "white")
```

#### Plot the Novelty after manual filtering 
```{r, fig.width=10, fig.height=20}
plot_novelty(sce_filter_manual, "additional UMI")
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter_novelty.", ini$file_type), 
device = ini$file_type, width = 10, height = 14, bg = "white")
```

#### Filter individual samples further as required


A few more manual filters, again this will be dependent on the data  
Remember to update the manual_filters df to record what settings were used  

```{r, fig.width=10, fig.height=16, message = FALSE}
plot_mito(sce_filter_manual)
```


```{r}
ggsave(paste0(directory, "/QC_figures/MAD_filter_mito_Vln.", ini$file_type), 
    device = ini$file_type, width = 10, height = 16, bg = "white")
```


Some further mitochondrial filtering and nFeatures filtering
MSK0785-Ach-muscle	4% mt
MSK1250-ACH-muscle	4% mt
MSK1556-Ach-Enth	4% mt
MSK1556-Ach-MTJ	4% mt
MSK1691-ACH-MTJ	4% mt


```{r}

sce_filter_manual[["MSK0785-Ach-muscle"]] <- 
    sce_filter_manual[["MSK0785-Ach-muscle"]][, sce_filter_manual[["MSK0785-Ach-muscle"]]$subsets_mito_percent < 4 ]
sce_filter_manual[["MSK1250-ACH-muscle"]] <- 
    sce_filter_manual[["MSK1250-ACH-muscle"]][, sce_filter_manual[["MSK1250-ACH-muscle"]]$subsets_mito_percent < 4 ]
sce_filter_manual[["MSK1556-Ach-Enth"]] <- 
    sce_filter_manual[["MSK1556-Ach-Enth"]][, sce_filter_manual[["MSK1556-Ach-Enth"]]$subsets_mito_percent < 4 ]
sce_filter_manual[["MSK1556-Ach-MTJ"]] <- 
    sce_filter_manual[["MSK1556-Ach-MTJ"]][, sce_filter_manual[["MSK1556-Ach-MTJ"]]$subsets_mito_percent < 4 ]
sce_filter_manual[["MSK1691-ACH-MTJ"]] <- 
    sce_filter_manual[["MSK1691-ACH-MTJ"]][, sce_filter_manual[["MSK1691-ACH-MTJ"]]$subsets_mito_percent < 4 ]



# update the manual_filters df  
manual_filters[manual_filters$sample =="MSK0785-Ach-muscle", "mito_filter"] <- 4
manual_filters[manual_filters$sample =="MSK1250-ACH-muscle", "mito_filter"] <- 4
manual_filters[manual_filters$sample =="MSK1556-Ach-Enth", "mito_filter"] <- 4
manual_filters[manual_filters$sample =="MSK1556-Ach-MTJ", "mito_filter"] <- 4
manual_filters[manual_filters$sample =="MSK1691-ACH-MTJ", "mito_filter"] <- 4

# save manual filters table
write.table (manual_filters, 
             file = paste0(directory, "/manual_filters.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
manual_filters

```

#### Plot the number of cells after final filtering

```{r, fig.width=10, fig.height = 6}
plot_cell_number(sce_filter_manual, "cell", "final")
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter2_cellnumber.", ini$file_type), 
       device = ini$file_type, width = 10, height = 6, bg = "white")
```

#### Plot the QC metrics after final filtering

```{r, fig.width=12, fig.height=14}
plot_filtered_qc(sce_filter_manual, "final")
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter2_Vlnplots.", ini$file_type), 
device = ini$file_type, width = 12, height = 14, bg = "white")
```

#### Plot UMI vs genes detected after final filtering

```{r, fig.width=14, fig.height=24, message = FALSE}
plot_qc(sce_filter_manual, "final")

```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter2_UMIvsgenes.", ini$file_type), 
device = ini$file_type, width = 14, height = 24, bg = "white")
```

#### Plot the relationship between nUMI and % MT after final filtering 

```{r, fig.width=10, fig.height=20}
p <- plot_umi_vs_mito(sce_filter_manual, "final")
p
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter2_nUMI_vs_mito.", ini$file_type), 
    device = ini$file_type, width = 14, height = 24, bg = "white")

```



#### Plot the Novelty after final filtering 
```{r, fig.width=10, fig.height=20}
plot_novelty(sce_filter_manual, "final")
```

```{r}
ggsave(paste0(directory, "/QC_figures/Manual_filter2_novelty.", ini$file_type), 
device = ini$file_type, width = 10, height = 12, bg = "white")
```
#### Save the filtered sce objects 

```{r, warning=FALSE}

dir.create(paste0(directory, "/RDS_objects.dir/filtered/"))

so <- list()

for (i in 1:length(sce_filter_manual)){
    
    saveRDS(sce_filter_manual[[i]], 
            paste0(directory, "/RDS_objects.dir/filtered/", mainExpName(sce[[i]]), "_filtered_SingleCellExp.rds"))
    
}

```

If required can also save as Seurat object.  
Change eval=TRUE if required. 

```{r, warning=FALSE, eval=FALSE}
so <- list()
for (i in 1:length(sce_filter_manual)){
    #remove "-" from expname
    mainExpName(sce_filter_manual[[i]]) <- mainExpName(sce_filter_manual[[i]]) %>% str_replace_all("-", ".")
    # convert the sce to a so
    # NB using project name here doesn't work
    so[[i]] <- as.Seurat(sce_filter_manual[[i]], 
                             data = NULL) #because there are no logcounts
    # the assay has the mainExpName
    
    # remove some extra columns from the metadata that are not needed
    so[[i]][[paste0("nCount_", mainExpName(sce_filter_manual[[i]]))]] <- NULL
    so[[i]][[paste0("nFeature_", mainExpName(sce_filter_manual[[i]]))]] <- NULL
    
    # add a project name
    so[[i]]@project.name <- mainExpName(sce_filter_manual[[i]])
    
    # save the seurat object
    saveRDS(so[[i]],
           paste0(directory, "/RDS_objects.dir/filtered/", 
                  mainExpName(sce_filter_manual[[i]]), "_SeuratObject.rds"))
   
}

so

```

#### Record the settings used

```{r, echo = FALSE}


cat("This is a description of the following run on ", date, "\n")
cat("\n")
cat("Run Location:\n")
cat(getwd(), "\n")
cat("\n")
cat("This run analysed the following samples:\n")
cat(names(sce), sep = "\n")
cat("\n")
cat("See the file manual_filters.txt for the filtering thresholds used. \n")


```




